### PR TITLE
[Cognitive Services] Upgrade to 2021-04-30

### PR DESCRIPTION
**Pull request by yangyuan**
## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - {Please put the link here}
    - [x] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by msJinLei** [comment]
@yangyuan
You commit includes a lot of breaking changes which cannot be merged to master.
If all the breaking changes are required for SDK upgrading, you have to wait until next major release.
if not, you can extract essential changes and covert them to unbreaking ones.

**Comment by yangyuan** [comment]
@msJinLei is next major release on Oct?

Most of the changes are related to SDK. Let me see if I can do something in here to avoid breaking changes.

Some of the changes are marked as breaking but it's actually bug fix, for example ActionRequired is removed from PrivateLinkServiceConnectionState.
The field ActionRequired as incorrectly defined in SDK, the real name should be ActionsRequired. This kind bug fix, can we include marge them without waiting for new major version?

**Comment by msJinLei** [comment]
> @msJinLei is next major release on Oct?
> 
> Most of the changes are related to SDK. Let me see if I can do something in here to avoid breaking changes.
> 
> Some of the changes are marked as breaking but it's actually bug fix, for example ActionRequired is removed from PrivateLinkServiceConnectionState.
> The field ActionRequired was incorrectly defined in SDK, we got breaking change approval a while ago for this bugfix, the real name should be ActionsRequired.

I understand your point. However, Azure PowerShell prevents any breaking changes being released in non-breaking change release.

```
  "AssemblyFileName","ClassName","Target","Severity","ProblemId","Description","Remediation"
  "Az.CognitiveServices","Microsoft.Azure.Commands.Management.CognitiveServices.GetAzureCognitiveServicesAccountCommand","Get-AzCognitiveServicesAccount","0","3030","The generic type for 'property Type' has been changed from 'System.Nullable`1[Microsoft.Azure.Management.CognitiveServices.Models.IdentityType]' to 'System.Nullable`1[Microsoft.Azure.Management.CognitiveServices.Models.ResourceIdentityType]'. ","Change the generic type for 'property Type' back to 'System.Nullable`1[Microsoft.Azure.Management.CognitiveServices.Models.IdentityType]'."
  "Az.CognitiveServices","Microsoft.Azure.Commands.Management.CognitiveServices.GetAzureCognitiveServicesAccountCommand","Get-AzCognitiveServicesAccount","0","3000","The type of property 'Tier' of type 'Microsoft.Azure.Management.CognitiveServices.Models.Sku' has changed from 'System.Nullable`1[Microsoft.Azure.Management.CognitiveServices.Models.SkuTier]' to 'System.String'.","Change the type of property 'Tier' back to 'System.Nullable`1[Microsoft.Azure.Management.CognitiveServices.Models.SkuTier]'."
  "Az.CognitiveServices","Microsoft.Azure.Commands.Management.CognitiveServices.GetAzureCognitiveServicesAccountCommand","Get-AzCognitiveServicesAccount","0","3010","The property 'ActionRequired' of type 'Microsoft.Azure.Management.CognitiveServices.Models.PrivateLinkServiceConnectionState' has been removed.","Add the property 'ActionRequired' back to type 'Microsoft.Azure.Management.CognitiveServices.Models.PrivateLinkServiceConnectionState'."
```

File: src/CognitiveServices/CognitiveServices/CognitiveServicesAccount/GetAzureCognitiveServicesAccount.cs [hunk 1]
```diff
@@ -53,46 +62,86 @@ public class GetAzureCognitiveServicesAccountCommand : CognitiveServicesAccountB
...
+
             });
         }
 
-        private IEnumerable<CognitiveServicesAccount> GetWithPaging(IPage<CognitiveServicesAccount> firstPage, bool isResourceGroup)
+        private IEnumerable<Account> GetWithPaging(IPage<Account> firstPage, bool isResourceGroup)
         {
-            var cognitiveServicesAccounts = new List<CognitiveServicesAccount>(firstPage);
-            IPage<CognitiveServicesAccount> nextPage = null;
+            var cognitiveServicesAccounts = new List<Account>(firstPage);
```
Comment (msJinLei) [inline src/CognitiveServices/CognitiveServices/CognitiveServicesAccount/GetAzureCognitiveServicesAccount.cs:143]: You should return the type the same as what you defined as cmdlet output.

**Review by msJinLei — CHANGES_REQUESTED** [review]

File: src/CognitiveServices/CognitiveServices/CognitiveServicesAccount/GetAzureCognitiveServicesAccount.cs (hunk 1 above)
Comment (yangyuan) [reply src/CognitiveServices/CognitiveServices/CognitiveServicesAccount/GetAzureCognitiveServicesAccount.cs:143]: Yes, `WriteCognitiveServicesAccount` and `WriteCognitiveServicesAccountList` will convert this to `PSCognitiveServicesAccount`

**Comment by yangyuan** [comment]
Hi @msJinLei , few things:

1. Can you check that if everything else are fine, except for the last 3 breaking changes.
2. Due to legal concern we might have to update PowerShell to latest RP version.
2.1. It seems to be impossible to fix the last few breaking changes with latest SDK. Is there any way to request exception? 
2.2. I can, also, copy the majority of SDK code to this repo, and modify it to make it compatible with old version.

**Comment by msJinLei** [comment]
> Hi @msJinLei , few things:
> 
> 1. Can you check that if everything else are fine, except for the last 3 breaking changes.
> 2. Due to legal concern we might have to update PowerShell to latest RP version.
>    2.1. It seems to be impossible to fix the last few breaking changes with latest SDK. Is there any way to request exception?
>    2.2. I can, also, copy the majority of SDK code to this repo, and modify it to make it compatible with old version. Thus we get rid of SDK dependency (temporary), and re-introduce it in next major release.

I check breaking change items one by one but cannot the following change.
````
"Az.CognitiveServices","Microsoft.Azure.Commands.Management.CognitiveServices.GetAzureCognitiveServicesAccountCommand","Get-AzCognitiveServicesAccount","0","3010","The property 'ActionRequired' of type 'Microsoft.Azure.Management.CognitiveServices.Models.PrivateLinkServiceConnectionState' has been removed."
````
I check the class definition. ActionRequired is not removed.
```csharp
#region Assembly Microsoft.Azure.Management.CognitiveServices, Version=8.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
// C:\Users\leijin\.nuget\packages\microsoft.azure.management.cognitiveservices\8.0.0-preview\lib\netstandard2.0\Microsoft.Azure.Management.CognitiveServices.dll
#endregion

using Newtonsoft.Json;

namespace Microsoft.Azure.Management.CognitiveServices.Models
{
    //
    // Summary:
    //     A collection of information about the state of the connection between service
    //     consumer and provider.
    public class PrivateLinkServiceConnectionState
    {
        //
        // Summary:
        //     Initializes a new instance of the PrivateLinkServiceConnectionState class.
        public PrivateLinkServiceConnectionState();
        //
        // Summary:
        //     Initializes a new instance of the PrivateLinkServiceConnectionState class.
        //
        // Parameters:
        //   status:
        //     Indicates whether the connection has been Approved/Rejected/Removed by the owner
        //     of the service. Possible values include: 'Pending', 'Approved', 'Rejected'
        //
        //   description:
        //     The reason for approval/rejection of the connection.
        //
        //   actionsRequired:
        //     A message indicating if changes on the service provider require any updates on
        //     the consumer.
        public PrivateLinkServiceConnectionState(string status = null, string description = null, string actionsRequired = null);

        //
        // Summary:
        //     Gets or sets indicates whether the connection has been Approved/Rejected/Removed
        //     by the owner of the service. Possible values include: 'Pending', 'Approved',
        //     'Rejected'
        [JsonProperty(PropertyName = "status")]
        public string Status { get; set; }
        //
        // Summary:
        //     Gets or sets the reason for approval/rejection of the connection.
        [JsonProperty(PropertyName = "description")]
        public string Description { get; set; }
        //
        // Summary:
        //     Gets or sets a message indicating if changes on the service provider require
        //     any updates on the consumer.
        [JsonProperty(PropertyName = "actionsRequired")]
        public string ActionsRequired { get; set; }
    }
}
```

File: src/CognitiveServices/CognitiveServices/help/Az.CognitiveServices.md [hunk 2]
```diff
@@ -47,6 +50,9 @@ Remove IpRules or VirtualNetworkRules from the NetWorkRule property of a Cogniti
 ### [Set-AzCognitiveServicesAccount](Set-AzCognitiveServicesAccount.md)
 Modifies an account.
 
+### [Undo-AzCognitiveServicesAccountRemoval](Undo-AzCognitiveServicesAccountRemoval.md)
+{{ Fill in the Synopsis }}
```
Comment (msJinLei) [inline src/CognitiveServices/CognitiveServices/help/Az.CognitiveServices.md:54]: Synopsis should not be empty

File: src/CognitiveServices/CognitiveServices/help/New-AzCognitiveServicesAccountApiProperty.md [hunk 3]
```diff
@@ -51,37 +49,6 @@ Accept pipeline input: False
 Accept wildcard characters: False
 ```
 
-### -Confirm
```
Comment (msJinLei) [inline src/CognitiveServices/CognitiveServices/help/New-AzCognitiveServicesAccountApiProperty.md:54]: whatif and confirm should not be removed

File: src/CognitiveServices/CognitiveServices/CognitiveServicesAccount/NewAzureCognitiveServicesAccountApiProperty.cs [hunk 4]
```diff
@@ -21,7 +21,7 @@ namespace Microsoft.Azure.Commands.Management.CognitiveServices
     /// <summary>
     /// Generate Cognitive Services Account ApiProperties class
     /// </summary>
-    [Cmdlet("New", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "CognitiveServicesAccountApiProperty"), OutputType(typeof(CognitiveServicesModels.CognitiveServicesAccountApiProperties))]
+    [Cmdlet("New", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "CognitiveServicesAccountApiProperty"), OutputType(typeof(CognitiveServicesModels.ApiProperties))]
```
Comment (msJinLei) [inline src/CognitiveServices/CognitiveServices/CognitiveServicesAccount/NewAzureCognitiveServicesAccountApiProperty.cs:24]: SupportsShouldProcess should be true

File: src/CognitiveServices/CognitiveServices/CognitiveServicesAccount/UndoAzureCognitiveServicesAccountRemoval.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+using Microsoft.Azure.Management.CognitiveServices.Models;
+using System.Globalization;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Management.CognitiveServices
+{
+    /// <summary>
+    /// Delete a Cognitive Services.
+    /// </summary>
+    [Cmdlet("Undo", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "CognitiveServicesAccountRemoval", SupportsShouldProcess = true), OutputType(typeof(void))]
```
Comment (msJinLei) [inline src/CognitiveServices/CognitiveServices/CognitiveServicesAccount/UndoAzureCognitiveServicesAccountRemoval.cs:27]: output type doesn't seem to be void

**Review by msJinLei — CHANGES_REQUESTED** [review]

**Comment by yangyuan** [comment]
@msJinLei `ActionRequired  `is removed, what you can see in PrivateLinkServiceConnectionState is `Action s Required` .

Thanks for feedback, I'm DRI this week but will get back to you as soon as I get time.

**Comment by msJinLei** [comment]
> @msJinLei `ActionRequired `is removed, what you can see in PrivateLinkServiceConnectionState is `Action s Required` .
> 
> Thanks for feedback, I'm DRI this week but will get back to you as soon as I get time.

**Comment by BethanyZhou** [comment]
why you can merge commit 855cfd5 into master branch?

**Pull request merged**